Core/src/org/sleuthkit/autopsy/casemodule/AddLogicalImageTask.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.commons.io.FileUtils;
+import org.openide.util.NbBundle.Messages;
+import org.sleuthkit.autopsy.corecomponentinterfaces.DataSourceProcessorCallback;
+import org.sleuthkit.autopsy.corecomponentinterfaces.DataSourceProcessorProgressMonitor;
+import org.sleuthkit.autopsy.coreutils.Logger;
+import org.sleuthkit.autopsy.imagewriter.ImageWriterSettings;
+import org.sleuthkit.datamodel.Content;
+import org.sleuthkit.datamodel.TskCoreException;
+
+public class AddLogicalImageTask extends AddImageTask {
+
+    private final static Logger logger = Logger.getLogger(AddLogicalImageTask.class.getName());   
+    private final static String ALERT_TXT = "alert.txt"; //NON-NLS
+    private final static String USERS_TXT = "users.txt"; //NON-NLS
+    private final File src;
+    private final File dest;
+    private final DataSourceProcessorCallback callback;
+    private final DataSourceProcessorProgressMonitor progressMonitor;
+    
+    public AddLogicalImageTask(String deviceId, String imagePath, int sectorSize, 
+            String timeZone, boolean ignoreFatOrphanFiles, 
+            String md5, String sha1, String sha256, 
+            ImageWriterSettings imageWriterSettings, 
+            File src, File dest,
+            DataSourceProcessorProgressMonitor progressMonitor, 
+            DataSourceProcessorCallback callback
+    ) {
+        super(deviceId, imagePath, sectorSize, timeZone, ignoreFatOrphanFiles, 
+                md5, sha1, sha256, imageWriterSettings, progressMonitor, callback);
+        this.src = src;
+        this.dest = dest;
+        this.progressMonitor = progressMonitor;
+        this.callback = callback;
+    }
+    
+    /**
+     * Copy the src directory to dest.
+     * Add alert.txt and users.txt to the case report
+     * Adds the image to the case database.
+     */
+    @Messages({
+        "# {0} - src", "# {1} - dest", "AddLogicalImageTask.copyingImageFromTo=Copying image from {0} to {1}",
+        "AddLogicalImageTask.doneCopying=Done copying",
+        "# {0} - src", "# {1} - dest", "AddLogicalImageTask.failedToCopyDirectory=Failed to copy directory {0} to {1}",
+        "# {0} - file", "AddLogicalImageTask.addingToReport=Adding {0} to report",
+        "# {0} - file", "AddLogicalImageTask.doneAddingToReport=Done adding {0} to report"
+    })
+    @Override
+    public void run() {
+        List<String> errorList = new ArrayList<>();
+        List<Content> emptyDataSources = new ArrayList<>();
+        
+        try {
+            progressMonitor.setProgressText(Bundle.AddLogicalImageTask_copyingImageFromTo(src.toString(), dest.toString()));
+            FileUtils.copyDirectory(src, dest);
+            progressMonitor.setProgressText(Bundle.AddLogicalImageTask_doneCopying());
+        } catch (IOException ex) {
+            // Copy directory failed
+            String msg = Bundle.AddLogicalImageTask_failedToCopyDirectory(src.toString(), dest.toString());
+            logger.log(Level.SEVERE, msg);
```

Comment:
Forgot to put this earlier. Our log messages should never be from the bundles - they should always just be a text string. This helps us diagnose problems through the logs. There are a couple of these to fix.

Reply:
fixed

Reply:
I think it's still not quite right. The messages added to errorList are displayed to the user and should be bundle messages, but the log messages should be strings, so you'll have two very similar strings in the bundle message and in the log message.

Reply:
I think we should not log the IOException from copyDirectory. There are all kinds of possible IOExceptions while copying: USB drive ejected, out of disk space etc. These are *not* program logic errors that Autopsy developer should be debug.

The user will get the error message in the wizard dialog box. 

The exception in addReport should be logged and reported to user. TskCoreException is from our code/logic so it is not obvious to the user what caused this exception.

I'll fix them again

Reply:
No, we need to log all exceptions. I can tell you from long experience that it can be very difficult to get error messages from the UI from a user, while it is often possible for the user to simply zip up the log files and send them to us. Also, without the stack traces and exceptions messages we lose valuable information. The coding guideline for Autopsy development is that we never simply swallow exceptions and ALL exceptions are logged, with stack traces. 

Guidelines for logging levels:
INFO - provide context for other log messages and duration info. For example, logging the start and finish of ingest jobs.
WARNING - an exception occurred, but the integrity of the case is not in question. Example, a corrupted file was encountered.
SEVERE - an exception occurred that makes the case incomplete. Example, failure to insert a record in the case database.